File: apps/go-modular/internal/config/validate.go
package config

import (
	"errors"
	"fmt"
	"net/url"
	"slices"
	"strings"
)

// Validates critical configuration values
func validateConfig(config *Config) error {
	if config == nil {
		return fmt.Errorf("config is nil")
	}

	var errs []string

	// App mode
	mode := strings.ToLower(strings.TrimSpace(config.App.Mode))
	if mode == "" {
		errs = append(errs, "app mode is required")
	} else {
		validModes := []string{"development", "production", "staging", "test"}
		if !slices.Contains(validModes, mode) {
			errs = append(errs, fmt.Sprintf("invalid app mode: %q (valid: %v)", config.App.Mode, validModes))
		}
	}

	// JWT algorithm and secret
	alg := strings.ToUpper(strings.TrimSpace(string(config.App.JWTAlgorithm)))
	if alg != "" {
		validAlgs := []string{string(JWTAlgorithmHS256), string(JWTAlgorithmRS256)}
		if !slices.Contains(validAlgs, alg) {
			errs = append(errs, fmt.Sprintf("invalid JWT algorithm: %q (valid: %v)", alg, validAlgs))
		}
	}
	secret := strings.TrimSpace(config.App.JWTSecretKey)
	if strings.EqualFold(mode, "production") {
		if secret == "" || secret == "_THIS_IS_DEFAULT_JWT_SECRET_KEY_" {
			errs = append(errs, "JWT secret key must be set in production")
		}
	}

	// Server port
	if config.App.ServerPort <= 0 || config.App.ServerPort > 65535 {
		errs = append(errs, fmt.Sprintf("invalid server port: %d (must be 1-65535)", config.App.ServerPort))
	}

	// Database URL
	dbURL := strings.TrimSpace(config.Database.PostgresURL)
	if dbURL == "" {
		errs = append(errs, "database URL is required")
	} else {
		if _, err := url.Parse(dbURL); err != nil {
			errs = append(errs, fmt.Sprintf("invalid database URL: %v", err))
		}
	}

	// Postgres pool size / retries
	if config.Database.PgMaxPoolSize <= 0 {
		errs = append(errs, fmt.Sprintf("pg max pool size must be > 0 (got %d)", config.Database.PgMaxPoolSize))
	}
	if config.Database.PgMaxRetries < 0 {
		errs = append(errs, fmt.Sprintf("pg max retries must be >= 0 (got %d)", config.Database.PgMaxRetries))
	}

	// Mailer
	mh := strings.TrimSpace(config.Mailer.SMTPHost)
	if mh != "" {
		if config.Mailer.SMTPPort <= 0 || config.Mailer.SMTPPort > 65535 {
			errs = append(errs, fmt.Sprintf("invalid mailer SMTP port: %d (must be 1-65535)", config.Mailer.SMTPPort))
		}
	} else {
		// If host is blank but port set, warn/error
		if config.Mailer.SMTPPort > 0 {
			errs = append(errs, "mailer SMTP host is empty but SMTP port is set")
		}
	}

	// File store / S3
	s3ep := strings.TrimSpace(config.FileStore.S3Endpoint)
	if s3ep != "" {
		if strings.TrimSpace(config.FileStore.S3BucketName) == "" {
			errs = append(errs, "file store S3 bucket name is required when S3 endpoint is set")
		}
		// In production require credentials
		if strings.EqualFold(mode, "production") {
			if config.FileStore.S3AccessKey == "" || config.FileStore.S3SecretKey == "" {
				errs = append(errs, "S3 access key and secret are required in production when S3 endpoint is set")
			}
		}
	}

	// Logging format & request body logging
	logFormat := strings.ToLower(strings.TrimSpace(config.Logging.Format))
	if logFormat != "" {
		validFormats := []string{"json", "pretty"}
		if !slices.Contains(validFormats, logFormat) {
			errs = append(errs, fmt.Sprintf("invalid log format: %q (valid: %v)", config.Logging.Format, validFormats))
		}
	}

	// Rate limiting
	if config.App.RateLimitEnabled {
		if config.App.RateLimitRequests <= 0 {
			errs = append(errs, "rate limit requests must be > 0 when rate limiting is enabled")
		}
		if config.App.RateLimitDuration <= 0 {
			errs = append(errs, "rate limit duration must be > 0 when rate limiting is enabled")
		}
	}

	// Telemetry
	if config.OTel.EnableTelemetry {
		if strings.TrimSpace(config.OTel.ExporterOTLPEndpoint) == "" {
			errs = append(errs, "OTel exporter endpoint is required when telemetry is enabled")
		}
		if config.OTel.TracingSampleRate < 0 || config.OTel.TracingSampleRate > 1 {
			errs = append(errs, fmt.Sprintf("OTel tracing sample rate must be between 0 and 1 (got %v)", config.OTel.TracingSampleRate))
		}
	}

	if len(errs) > 0 {
		return errors.New(strings.Join(errs, "; "))
	}
	return nil
}

File: templates/rust-modular/src/config/validate.rs
//! Config validation — port of `internal/config/validate.go` plus two
//! Phase D extras:
//!
//! 1. Reject plaintext SMTP ports (25, 1025) in production mode.
//!    Complements the design 3.5 deletion of `SMTPSecure`: the
//!    lettre transport builder enforces TLS on 587/465; plaintext
//!    is only allowed in dev.
//!
//! 2. No RS256 / `JWT_ALGORITHM` validation (design 3.4: HS256 only).

use anyhow::{Result, bail};

use super::types::Config;

/// Validate the fully-loaded config. Returns a joined error with all
/// failing checks so operators see every problem in one boot attempt.
pub fn validate_config(config: &Config) -> Result<()> {
    let mut errs: Vec<String> = Vec::new();

    // App mode
    let mode = config.app.app_mode.trim().to_ascii_lowercase();
    if mode.is_empty() {
        errs.push("app mode is required".to_string());
    } else if !matches!(
        mode.as_str(),
        "development" | "production" | "staging" | "test"
    ) {
        errs.push(format!(
            "invalid app mode: \"{}\" (valid: development, production, staging, test)",
            config.app.app_mode
        ));
    }

    // JWT secret
    let secret = config.app.jwt_secret_key.trim();
    if mode == "production" && (secret.is_empty() || secret == "_THIS_IS_DEFAULT_JWT_SECRET_KEY_") {
        errs.push("JWT secret key must be set in production".to_string());
    }

    // Server port
    if config.app.server_port == 0 {
        errs.push(format!(
            "invalid server port: {} (must be 1-65535)",
            config.app.server_port
        ));
    }

    // Database URL — loose check. sqlx will return a clearer error on
    // actual connect; validator only catches empty + obviously wrong
    // scheme. Matches Go's `url.Parse` which was very permissive.
    let db_url = config.database.database_url.trim();
    if db_url.is_empty() {
        errs.push("database URL is required".to_string());
    } else if !(db_url.starts_with("postgres://") || db_url.starts_with("postgresql://")) {
        errs.push(format!(
            "invalid database URL: \"{db_url}\" (must start with postgres:// or postgresql://)"
        ));
    }

    // Postgres pool
    if config.database.pg_max_pool_size == 0 {
        errs.push(format!(
            "pg max pool size must be > 0 (got {})",
            config.database.pg_max_pool_size
        ));
    }
    if config.database.pg_max_retries < -1 {
        errs.push(format!(
            "pg max retries must be >= -1 (got {}); -1 means infinite retry",
            config.database.pg_max_retries
        ));
    }

    // Mailer host/port
    let smtp_host = config.mailer.smtp_host.trim();
    if !smtp_host.is_empty() && config.mailer.smtp_port == 0 {
        errs.push(format!(
            "invalid mailer SMTP port: {} (must be 1-65535)",
            config.mailer.smtp_port
        ));
    }

    // Phase D extra: reject non-TLS SMTP ports in production.
    // Ports 587 (STARTTLS) and 465 (implicit TLS) are the only allowed
    // ports in production mode. Anything else — including the common
    // 25/1025 dev plaintext ports — is rejected to surface
    // misconfiguration at boot rather than silently sending plaintext.
    if mode == "production" && !smtp_host.is_empty() {
        let port = config.mailer.smtp_port;
        if port != 587 && port != 465 {
            errs.push(format!(
                "SMTP port {port} forbidden in production; use 587 (STARTTLS) or 465 (implicit TLS)"
            ));
        }
    }

    // FileStore
    let s3_ep = config.file_store.s3_endpoint.trim();
    if !s3_ep.is_empty() {
        if config.file_store.s3_bucket_name.trim().is_empty() {
            errs.push("file store S3 bucket name is required when S3 endpoint is set".to_string());
        }
        if mode == "production"
            && (config.file_store.s3_access_key.is_empty()
                || config.file_store.s3_secret_key.is_empty())
        {
            errs.push(
                "S3 access key and secret are required in production when S3 endpoint is set"
                    .to_string(),
            );
        }
    }

    // Logging format
    let log_format = config.logging.log_format.trim().to_ascii_lowercase();
    if !log_format.is_empty() && !matches!(log_format.as_str(), "json" | "pretty") {
        errs.push(format!(
            "invalid log format: \"{}\" (valid: json, pretty)",
            config.logging.log_format
        ));
    }

    // Rate limiting
    if config.app.rate_limit_enabled {
        if config.app.rate_limit_requests == 0 {
            errs.push("rate limit requests must be > 0 when rate limiting is enabled".to_string());
        }
        if config.app.rate_limit_burst_size == 0 {
            errs.push(
                "rate limit burst size must be > 0 when rate limiting is enabled".to_string(),
            );
        }
    }

    // OTel
    if config.otel.otel_enable_telemetry {
        if config.otel.otel_exporter_otlp_endpoint.trim().is_empty() {
            errs.push("OTel exporter endpoint is required when telemetry is enabled".to_string());
        }
        if !(0.0..=1.0).contains(&config.otel.otel_tracing_sample_rate) {
            errs.push(format!(
                "OTel tracing sample rate must be between 0 and 1 (got {})",
                config.otel.otel_tracing_sample_rate
            ));
        }
    }

    if !errs.is_empty() {
        bail!("config validation failed: {}", errs.join("; "));
    }
    Ok(())
}
